lib/Dialect/FIRRTL/Export/FIREmitter.cpp

```diff
@@ -0,0 +1,714 @@
+  emitLocationAndNewLine(op);
+}
+
+void Emitter::emitExpression(Value value) {
+  // Handle the trivial case where we already have a name for this value which
+  // we can use.
+  if (auto name = lookupEmittedName(value)) {
+    os << name;
+    return;
+  }
+
+  auto op = value.getDefiningOp();
+  assert(op && "value must either be a block arg or the result of an op");
+  TypeSwitch<Operation *>(op)
+      .Case<
+          // Basic expressions
+          ConstantOp, SpecialConstantOp, SubfieldOp, SubindexOp, SubaccessOp,
+          // Binary
+          AddPrimOp, SubPrimOp, MulPrimOp, DivPrimOp, RemPrimOp, AndPrimOp,
+          OrPrimOp, XorPrimOp, LEQPrimOp, LTPrimOp, GEQPrimOp, GTPrimOp,
+          EQPrimOp, NEQPrimOp, CatPrimOp, DShlPrimOp, DShlwPrimOp, DShrPrimOp,
+          // Unary
+          AsSIntPrimOp, AsUIntPrimOp, AsAsyncResetPrimOp, AsClockPrimOp,
+          CvtPrimOp, NegPrimOp, NotPrimOp, AndRPrimOp, OrRPrimOp, XorRPrimOp,
+          // Miscellaneous
+          BitsPrimOp, HeadPrimOp, TailPrimOp, PadPrimOp, MuxPrimOp, ShlPrimOp,
+          ShrPrimOp>([&](auto op) { emitExpression(op); })
+      .Default([&](auto op) {
+        emitOpError(op, "not supported as expression");
+        os << "<unsupported-expr-" << op->getName().stripDialect() << ">";
+      });
+}
+
+void Emitter::emitExpression(ConstantOp op) {
+  emitType(op.getType());
+  // TODO: Add option to control base-2/8/10/16 output here.
+  os << "(" << op.value() << ")";
+}
+
+void Emitter::emitExpression(SpecialConstantOp op) {
+  auto emitInner = [&]() { os << "UInt<1>(" << op.value() << ")"; };
+  TypeSwitch<FIRRTLType>(op.getType().cast<FIRRTLType>())
+      .Case<ClockType>([&](auto type) {
+        os << "asClock(";
+        emitInner();
+        os << ")";
+      })
+      .Case<ResetType>([&](auto type) { emitInner(); })
+      .Case<AsyncResetType>([&](auto type) {
+        os << "asAsyncReset(";
+        emitInner();
+        os << ")";
+      });
+}
+
+void Emitter::emitExpression(SubfieldOp op) {
+  auto type = op.input().getType().cast<BundleType>();
+  emitExpression(op.input());
+  os << "." << type.getElementName(op.fieldIndex());
+}
+
+void Emitter::emitExpression(SubindexOp op) {
+  emitExpression(op.input());
+  os << "[" << op.index() << "]";
+}
+
+void Emitter::emitExpression(SubaccessOp op) {
+  emitExpression(op.input());
+  os << "[";
+  emitExpression(op.index());
+  os << "]";
+}
+
+void Emitter::emitBinaryExpression(Operation *op, StringRef mnemonic) {
+  os << mnemonic << "(";
+  emitExpression(op->getOperand(0));
+  os << ", ";
+  emitExpression(op->getOperand(1));
+  os << ")";
+}
+
+void Emitter::emitUnaryExpression(Operation *op, StringRef mnemonic) {
+  os << mnemonic << "(";
+  emitExpression(op->getOperand(0));
+  os << ")";
+}
+
+void Emitter::emitExpression(BitsPrimOp op) {
+  os << "bits(";
+  emitExpression(op.input());
+  os << ", " << op.hi() << ", " << op.lo() << ")";
+}
+
+void Emitter::emitExpression(HeadPrimOp op) {
```

Comment:
I like how you handled all the binary ops up above to eliminate the need for out of line implementations.  Maybe you could some overloaded methods like:
`void emitPrim(const char *, Value, Attribute, Attribute) ` and `void emitPrim(const char *, Value, Attribute)` etc, and then you could define each of these inline in the class as one liners like :

`void emitExpression(BitsPrimOp op) {  emitPrimitive("bits", op.input(), op.hi(), op.lo());}`

Reply:
or even `emitPrim(const char *name, Operation *op, Attr, Optional<Attr> = None)` and pull the value's out of the operand list of the operation.

Reply:
This was a very nice suggestion, cleaned things up an additional bunch. Pretty comfortable that the correspondence between the FIRRTL dialect and FIR syntax is so close.